Platform/AMD/AmdPlatformPkg/DynamicTables/Library/Acpi/AcpiSsdtPciLib/AcpiSsdtPciLib.c

```diff
@@ -0,0 +1,1540 @@
+    if (EFI_ERROR (Status)) {
+      ASSERT_EFI_ERROR (Status);
+      return Status;
+    }
+  }
+
+  if (EFI_ERROR (Status)) {
+    DEBUG ((
+      DEBUG_INFO,
+      "%a: Failed with Status: %r, Not Critical return SUCCESS\n",
+      __func__,
+      Status
+      ));
+    Status = EFI_SUCCESS;
+  }
+
+  return Status;
+}
+
+/**
+  Insert Pcie base size into the AML table
+
+  @param[in,out]  CrsNode    - AmlLib table node
+
+  @retval         EFI_SUCCESS, various EFI FAILUREs.
+**/
+EFI_STATUS
+EFIAPI
+InternalInsertPciExpressBaseSize (
+  IN OUT  AML_OBJECT_NODE_HANDLE  CrsNode
+  )
+{
+  EFI_STATUS  Status;
+  UINT64      RangeLen;
+  UINT64      RangeMax;
+  UINT64      RangeMin;
+
+  RangeMin = PcdGet64 (PcdPciExpressBaseAddress);
+  RangeLen = PcdGet64 (PcdPciExpressBaseSize);
+  RangeMax = RangeMin + RangeLen - 1;
+
+  Status = AmlCodeGenRdQWordMemory (
+             FALSE,
+             TRUE,
+             TRUE,
+             TRUE,
+             FALSE, // non cacheable
+             TRUE,
+             0x0,
+             RangeMin,
+             RangeMax,
+             0x0,
+             RangeLen,
+             0,
+             NULL,
+             0,
+             TRUE,
+             CrsNode,
+             NULL
+             );
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Locates an existing ACPI Table
+
+  @param[in]      Signature     - The Acpi table signature
+  @param[in]      OemTableId    - The Acpi table OEM Table ID. Ignored if 0
+  @param[out]     Table         - Table if Found or NULL
+
+  @retval         EFI_SUCCESS, various EFI FAILUREs.
+**/
+EFI_STATUS
+EFIAPI
+GetExistingAcpiTable (
+  IN      UINT32            Signature,
+  IN      UINT64            OemTableId,
+  OUT  EFI_ACPI_SDT_HEADER  **Table
+  )
+{
+  EFI_ACPI_SDT_HEADER     *LocalTable;
+  EFI_ACPI_TABLE_VERSION  LocalVersion;
+  EFI_STATUS              Status;
+  UINTN                   Index;
+  UINTN                   LocalTableKey;
+
+  Status = EFI_NOT_FOUND;
+  *Table = NULL;
+
+  for (Index = 0; ; Index++) {
+    Status = mAcpiSdtProtocol->GetAcpiTable (Index, &LocalTable, &LocalVersion, &LocalTableKey);
+    if (EFI_ERROR (Status)) {
+      return Status;
+    }
+
+    if (!(LocalTable->Signature == Signature)) {
+      continue;
+    }
+
+    // Accept table if OemTableId is zero.
+    if ((OemTableId == 0) ||
+        (CompareMem (&LocalTable->OemTableId, &OemTableId, 8) == 0))
+    {
+      *Table = LocalTable;
+      return EFI_SUCCESS;
+    }
+  }
+}
+
+/**
+  Install PCI devices scoped under \_SB into DSDT
+
+  Determine all the PCI Root Bridges and PCI root ports and install resources
+  including needed _HID, _CID, _UID, _ADR, _CRS and _PRT Nodes.
+
+  @param[in]      ImageHandle   - Standard UEFI entry point Image Handle
+  @param[in]      SystemTable   - Standard UEFI entry point System Table
+
+  @retval         EFI_SUCCESS, various EFI FAILUREs.
+**/
+EFI_STATUS
+GenerateAcpiSsdtPciTable (

```

Comment:
Do we have to clean up the memory or device node in the file conditions?

Reply:
done, added the exit_handler